File: Server/src/commonStructures/position.ts
/**
 * A 2d point in the space used to represent a position
 */
export default class Position {
    
    /**
     * The x coordinate
     */
    public x : number;
    
    /**
     * The y coordinate
     */
    public y : number;
    
    /**
     * Create a new position from the given coordinates
     * @param x The x coordinate
     * @param y The y coordinate
     */
    constructor(x:number, y:number) {
        this.x = x;
        this.y = y;
    }
    
    /**
     * Is the other object equal ?
     * @param other The other object to compare value
     * @returns Is the object values equal ?
     */
    public equals(other:Position):boolean{
        return other.x === this.x && other.y === this.y;
    }

    /**
     * Translates the position in the given direction
     * @param directionInDegree The direction in degrees
     * @param distance The distance to translate
     */
    public translate(directionInDegree:number, distance:number) {
        const directionRadiants = directionInDegree * Math.PI / 180;
        this.x += Math.cos(directionRadiants) * distance;
        this.y += Math.sin(directionRadiants) * distance;
    }

    /**
     * Add the other position to this one
     * @param other The other position to add
     */
    public add(other:Position){
        this.x += other.x;
        this.y += other.y;
    }

    /**
     * Subtract the other position to this one
     * @param other The other position to subtract
     */
    public subtract(other: Position) {
        this.x -= other.x;
        this.y -= other.y;
    }

    /**
     * Get a clone of this position (other object)
     * @returns A clone of this position
     */
    public clone():Position{
        return new Position(this.x,this.y);
    }

    /**
     * Get the distance between two positions
     * @param one The first position
     * @param two The second position
     * @returns The distance between the two positions
     */
    public static distance(one:Position,two:Position):number{
        return Math.sqrt(Math.pow(one.x - two.x,2) + Math.pow(one.y - two.y,2));
    }
}

File: Assets/Scripts/Network/CommonStructures/Position.cs
using UnityEngine;
using System;

namespace Network
{
    /// <summary>
    /// A 2D point used for position
    /// </summary>
    public class Position:IEquatable<Position>
    {
        /// <summary>
        /// The x coordinate
        /// </summary>
        public float x;

        /// <summary>
        /// The y coordinate
        /// </summary>
        public float y;

        /// <summary>
        /// Create a 0,0 position
        /// </summary>
        public Position()
        {
            this.x = 0;
            this.y = 0;
        }

        /// <summary>
        /// Create a new position with the given values
        /// </summary>
        /// <param name="x">The x axis position</param>
        /// <param name="y">The y axis position</param>
        public Position(float x, float y)
        {
            this.x = x;
            this.y = y;
        }

        /// <summary>
        /// Create a position from a Unity Vector2 (ONLY FOR CLIENT)
        /// </summary>
        /// <param name="vector2">The vector2</param>
        public Position(Vector2 vector2)
        {
            this.x = vector2.x;
            this.y = vector2.y;
        }

        /// <summary>
        /// Translate this position with an angle and a distance
        /// </summary>
        /// <param name="directionInDegree">The direction angle in degree</param>
        /// <param name="distance">The distance to translate in the given direction</param>
        public void Translate(float directionInDegree, float distance)
        {
            float radian = directionInDegree * Mathf.Deg2Rad;
            this.x += Mathf.Cos(radian) * distance;
            this.y += Mathf.Sin(radian) * distance;
        }

        /// <summary>
        /// Convert this object to a Unity Vector2 (ONLY FOR CLIENT)
        /// </summary>
        /// <returns>The Unity Vector2</returns>
        public Vector2 ToVector2()
        {
            return new Vector2(x, y);
        }

        /// <summary>
        /// Is the other object equal ?
        /// </summary>
        /// <param name="other">The position to compare</param>
        /// <returns>True if the same position</returns>
        public bool Equals(Position other)
        {
            return other.x == this.x && other.y == this.y;
        }
    }
}